## ENUNCIADO

En este caso se pretende obtener datos en la red social youtube.

En primer lugar, tenemos que darnos de alta como developers en la web google para acceder a sus herramientas 
(ver documento Ejemplos_Obtención_Acceso_APIs.pdf).
Para ejecutar los comandos del fichero debemos instalar el siguiente paquete en Docker:

`$ pip install google-api-python-client google-auth-httplib2 google-auth-oauthlib`

`$ pip install textblob`

Los apartados de la tarea son:

1. Los videos más vistos, por ejemplo 5 primeros, de una temática concreta, por
ejemplo, la diabetes, el cambio climático u otro que consideres oportuno.
Puedes elegir una temática o bien, pasar por teclado (ojo con el inglés). Usaremos
un idioma concreto como el castellano.

In [9]:
from googleapiclient.discovery import build

# Configuración de la API
API_KEY = "" # Clave API aqui 
YOUTUBE_API_SERVICE_NAME = "youtube"
YOUTUBE_API_VERSION = "v3"

# Temática del video
TEMA = "Iago Aspas"

# Construye y devuelve el cliente de la API de YouTube
def construir_cliente_youtube():
    return build(YOUTUBE_API_SERVICE_NAME, YOUTUBE_API_VERSION, developerKey=API_KEY)

# Obtiene una lista de videos según qué filtro apliquemos (más vistos o mejor valorados)
def obtener_videos(tema, max_resultados=5, idioma="es", orden="viewCount"):
    youtube = construir_cliente_youtube()

    # Realizar la búsqueda
    respuesta = youtube.search().list(
        q=tema,
        part="snippet",
        type="video",
        maxResults=max_resultados,
        relevanceLanguage=idioma,
        order=orden
    ).execute()

    # Procesar los resultados
    videos = [
        {
            "titulo": item["snippet"]["title"],
            "canal": item["snippet"]["channelTitle"],
            "video_id": item["id"]["videoId"]
        }
        for item in respuesta.get("items", [])
    ]
    return videos

# Muestra resultados de la búsqueda
def mostrar_videos(videos, filtro):
    print(f"\nVideos más {filtro} sobre '{TEMA}':")
    print("-" * 50)
    for i, video in enumerate(videos, start=1):
        print(f"Video {i}:")
        print(f"Título: {video['titulo']}")
        print(f"Canal: {video['canal']}")
        print(f"URL: https://www.youtube.com/watch?v={video['video_id']}")
        print("-" * 50)

if __name__ == "__main__":
    max_videos = 5
    idioma = "es"

    # Obtener y mostrar los videos más vistos
    mas_vistos = obtener_videos(TEMA, max_videos, idioma, orden="viewCount")
    mostrar_videos(mas_vistos, "vistos")



Videos más vistos sobre 'Iago Aspas':
--------------------------------------------------
Video 1:
Título: LA RESISTENCIA - Entrevista a Iago Aspas | #LaResistencia 23.10.2019
Canal: La Resistencia por Movistar Plus+
URL: https://www.youtube.com/watch?v=WWJtZSaSd7M
--------------------------------------------------
Video 2:
Título: IAGO ASPAS VS DELANTERO09 - Retos de Fútbol
Canal: Delantero09
URL: https://www.youtube.com/watch?v=e6n0ABHtFbI
--------------------------------------------------
Video 3:
Título: ¡¡¡ IAGO ASPAS vs TOBBAL !!! - El reto MAS ÉPICO de FUTBOL
Canal: TOBBALink
URL: https://www.youtube.com/watch?v=RKB1FtBrhbY
--------------------------------------------------
Video 4:
Título: CELTA DE VIGO 2 vs 2 FC BARCELONA | LALIGA 2024/25 MD14
Canal: FC Barcelona
URL: https://www.youtube.com/watch?v=VhWQPrmp7rU
--------------------------------------------------
Video 5:
Título: Iago ASPAS - Goals &amp; Assists - New Liverpool #9 - 2012/2013 - HD
Canal: YRCHDWorldwide
URL: http

2. Compara este resultado con el del video mejor valorado en esa temática. A veces
no coinciden.

In [5]:
# Obtener y mostrar los videos más valorados
mas_valorados = obtener_videos(TEMA, max_videos, "es", orden="rating")
mostrar_videos(mas_valorados, "valorados")


Videos más valorados sobre 'Iago Aspas':
--------------------------------------------------
Video 1:
Título: GOLAZO DE TIRO LIBRE IAGO ASPAS ⚽ Celta Vigo 1-0 FC Barcelona 🇪🇸
Canal: Gabriel Juárez
URL: https://www.youtube.com/watch?v=WOaeuvvuI-w
--------------------------------------------------
Video 2:
Título: MENSAXE IAGO ASPAS E LUCAS VÁZQUEZ GALA DO FÚTBOL GALEGO 18
Canal: Real Federación Galega de Fútbol
URL: https://www.youtube.com/watch?v=uYDX9pPWR6c
--------------------------------------------------
Video 3:
Título: Isco y Iago Aspas se chocan durante el partido ESPAÑA-RUSIA
Canal: Informacion Mundial
URL: https://www.youtube.com/watch?v=GYP0Rvh_YIE
--------------------------------------------------
Video 4:
Título: iago aspas hit the post before Henry 😂 #funk #fcmobile #bateson87
Canal: LC5
URL: https://www.youtube.com/watch?v=htSN-xlJCdw
--------------------------------------------------
Video 5:
Título: RAJADA de Iago Aspas CONTRA el Árbitro del Celta de Vigo vs Real Madrid

`Efectivamente, en este caso, no coinciden los vídeos más vistos y los más valorados.`

3. La librería TextBlob pertenece al API de Google, permite hacer análisis de
sentimientos con Python, entre otras funciones. Vas a investigar sobre esta librería
para resolver el siguiente ejercicio que consiste en obtener comentarios de un video
específico a tu elección y realizar un análisis de sentimientos en esos comentarios.
Obviamente tiene que tener comentarios. Debes imprimir los siguientes resultados:

    - print(f'Título del vídeo: {video_title}')
    - print(f'Número total de comentarios: {len(comments)}')
    - print(f'Promedio de sentimientos: {average_sentiment:.2f}')

In [12]:
from googleapiclient.discovery import build
from textblob import TextBlob

# Configuración de la API de YouTube
API_KEY = "AIzaSyArvf07AmR4Fa7LhsNBp44kYNcxQs-G_zQ"  # Clave de la API
YOUTUBE_API_SERVICE_NAME = "youtube"
YOUTUBE_API_VERSION = "v3"

# ID del video a analizar (extraído de la URL)
VIDEO_URL = "https://www.youtube.com/watch?v=kM-94LhhQTs"
VIDEO_ID = VIDEO_URL.split("v=")[-1]

def obtener_titulo_video(video_id):
    # Construcción del cliente de la API de YouTube
    youtube = build(YOUTUBE_API_SERVICE_NAME, YOUTUBE_API_VERSION, developerKey=API_KEY)
    
    # Llamada a la API para obtener el título del video
    respuesta = youtube.videos().list(
        part="snippet",
        id=video_id
    ).execute()
    
    # Extraer el título
    if respuesta["items"]:
        return respuesta["items"][0]["snippet"]["title"]
    else:
        return "Título no disponible"

def obtener_comentarios(video_id, max_resultados=100):
    # Construcción del cliente de la API de YouTube
    youtube = build(YOUTUBE_API_SERVICE_NAME, YOUTUBE_API_VERSION, developerKey=API_KEY)

    # Llamada a la API para obtener los comentarios
    respuesta = youtube.commentThreads().list(
        part="snippet",
        videoId=video_id,
        maxResults=max_resultados,
        textFormat="plainText"
    ).execute()

    comentarios = []
    for item in respuesta.get("items", []):
        comentario = item["snippet"]["topLevelComment"]["snippet"]["textDisplay"]
        comentarios.append(comentario)

    return comentarios

def analizar_sentimientos(comentarios):
    total_sentimiento = 0
    for comentario in comentarios:
        analisis = TextBlob(comentario)
        total_sentimiento += analisis.sentiment.polarity
    promedio_sentimiento = total_sentimiento / len(comentarios) if comentarios else 0
    return promedio_sentimiento

if __name__ == "__main__":
    print("Análisis de comentarios del video:")
    
    # Obtener título del video
    titulo_video = obtener_titulo_video(VIDEO_ID)
    
    # Obtener comentarios del video
    comentarios = obtener_comentarios(VIDEO_ID)
    
    if not comentarios:
        print("No se encontraron comentarios para este video.")
    else:
        # Realizar análisis de sentimientos
        promedio_sentimientos = analizar_sentimientos(comentarios)
        
        # Mostrar resultados
        print(f"Título del video: {titulo_video}")
        print(f"Número total de comentarios: {len(comentarios)}")
        print(f"Promedio de sentimientos: {promedio_sentimientos:.2f}")


Análisis de comentarios del video:
Título del video: Coldplay - Fix You (Glastonbury 2024)
Número total de comentarios: 100
Promedio de sentimientos: 0.20


`En este caso, empleando una muestra de 100 comentarios nos devuelve que en un rango de puntuación entre -1 y 1, el promedio de sentimientos para este video es de 0.20.
Esto quiere decir que los comentarios son de un tono neutral tirando a positivo.`